pkg/datapath/linux/routing/migrate.go

```diff
@@ -0,0 +1,242 @@
+	"github.com/cilium/cilium/pkg/option"
+
+	"github.com/vishvananda/netlink"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// MigrateENIRoutingTable migrates the egress rules inside the Linux routing
+// policy database (rpdb) for ENI IPAM mode.
+//
+// We rely on the ability to fetch the CiliumNode resource because we need to
+// fetch the number associated with the ENI device. The CiliumNode resource
+// contains this information in the Status field.
+//
+// This function should be invoked before exposing the Cilium API because we
+// want to ensure that no workloads are scheduled while this modification is
+// taking place. This migration is related to a bug
+// (https://github.com/cilium/cilium/issues/14336) where an ENI has an ifindex
+// that equals the main routing table number (253-255), causing the rules and
+// routes to be created using the wrong table ID, which could end up
+// blackholing most traffic on the node.
+func MigrateENIRoutingTable(getter ciliumNodeGetter) error {
+	if option.Config.IPAM != option.IPAMENI {
+		return nil
+	}
+
+	rules, err := netlink.RuleList(netlink.FAMILY_V4)
+	if err != nil {
+		return err
+	}
+
+	for _, r := range rules {
+		switch r.Priority {
+		case linux_defaults.RulePriorityEgress:
+			routes, err := netlink.RouteListFiltered(netlink.FAMILY_V4, &netlink.Route{
+				Table: r.Table,
+			}, netlink.RT_FILTER_TABLE)
+			if err != nil {
+				return err
+			}
+
+			// If there are no routes under the same table as the rule, then
+			// skip.
+			if len(routes) == 0 {
+				continue
+			}
+
+			oldTable := r.Table
+
+			// It is sufficient to grab the first route that matches because we
+			// are assuming all routes created under a rule will have the same
+			// ifindex (LinkIndex).
+			newTable, err := RetrieveTableIDFromIfIndex(routes[0].LinkIndex, getter)
+			if err != nil {
+				return err
+			}
+
+			if err := migrateOnUpgrade(oldTable, newTable); err != nil {
+				return err
+			}
+
+		case linux_defaults.RulePriorityEgressv2:
+			// Downgrade case. Unsupported for now.
+		}
+	}
+
+	return nil
+}
+
+func RetrieveTableIDFromIfIndex(ifindex int, getter ciliumNodeGetter) (int, error) {
+	link, err := netlink.LinkByIndex(ifindex)
+	if err != nil {
+		return -1, err
+	}
+
+	cn, err := getter.Get("", v1.GetOptions{})
+	if err != nil {
+		return -1, err
+	}
+
+	mac := link.Attrs().HardwareAddr
+
+	var eni v2.ENI
+	for _, e := range cn.Status.ENI.ENIs {
+		if e.MAC == mac.String() {
+			eni = e
+			break
+		}
+	}
+
+	return linux_defaults.RouteTableInterfacesOffset + eni.Number, nil
+}
+
+// migrateOnUpgrade migrates the ENI routing table for the upgrade path. It
+// accepts the old table ID for the existing rules to migrate. `new` is the new
+// table ID that the rules should be referring to.
+func migrateOnUpgrade(old, new int) error {
+	if old == new {
+		// Nothing to do.
+		return nil
+	}
+
+	if err := replaceDefaultRoutes(old, new); err != nil {
```

Comment:
`replace` here suggests that the routes in the `from` table will disappear and equivalent ones will appear in the `to` table. But if I read correctly, what is actually happening is we iterate the `from` table and then perform what is effectively an "Upsert" into the new table, potentially replacing existing routes in the destination table (that's what the "Replace" part in `RouteReplace()` is about, `unix.NLM_F_REPLACE` is set so the operation will replace any conflicting routes).

I also note the use of the words "Default routes" which has a well-known meaning in networking, specifically regarding the route for `0.0.0.0/0` (which matches all traffic). But from the function it appears to be copying all routes regardless of the prefix.

Super minor nit but if I'm following the above correctly then for readability I wonder if it would be clearer to reflect that in the naming of this function somehow, ie

```suggestion
	if err := copyRoutes(old, new); err != nil {
```

Reply:
That's a good point. I like your suggestion better.